components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -0,0 +1,336 @@
+      largest_block = log.block_number;
+    }
+  }
+  std::vector<mojom::BlockchainTokenPtr> discovered_assets;
+
+  for (const auto& contract_address : matching_contract_addresses) {
+    if (!tokens_to_search.contains(contract_address)) {
+      continue;
+    }
+    mojom::BlockchainTokenPtr token =
+        std::move(tokens_to_search.at(contract_address));
+
+    if (!BraveWalletService::AddUserAsset(token.Clone(), prefs_)) {
+      continue;
+    }
+    discovered_assets.push_back(std::move(token));
+  }
+
+  if (!triggered_by_accounts_added) {
+    DictionaryPrefUpdate update(prefs_,
+                                kBraveWalletNextAssetDiscoveryFromBlocks);
+    auto* next_asset_discovery_from_blocks = update.Get()->GetIfDict();
+    DCHECK(next_asset_discovery_from_blocks);
+    const auto path = base::StrCat({kEthereumPrefKey, ".", chain_id});
+    const std::string* current =
+        next_asset_discovery_from_blocks->FindStringByDottedPath(path);
+    uint256_t current_int = 0;
+    if (current) {
+      if (!HexValueToUint256(*current, &current_int)) {
+        CompleteDiscoverAssets(
+            chain_id, std::vector<mojom::BlockchainTokenPtr>(),
+            mojom::ProviderError::kInternalError,
+            l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR),
+            triggered_by_accounts_added);
+        return;
+      }
+    }
+    if ((!current || current_int <= largest_block) && largest_block > 0) {
+      next_asset_discovery_from_blocks->SetByDottedPath(
+          path, Uint256ValueToHex(largest_block + 1));
+    }
+  }
+  CompleteDiscoverAssets(chain_id, std::move(discovered_assets),
+                         mojom::ProviderError::kSuccess, "",
+                         triggered_by_accounts_added);
+}
+
+void AssetDiscoveryManager::CompleteDiscoverAssets(
+    const std::string& chain_id,
+    std::vector<mojom::BlockchainTokenPtr> discovered_assets_for_chain,
+    mojom::ProviderError error,
+    const std::string& error_message,
+    bool triggered_by_accounts_added,
+    DiscoverAssetsCompletedCallback callback) {
+  // Run callback (for tests only)
+  std::vector<mojom::BlockchainTokenPtr> discovered_assets_for_chain_clone;
+  for (const auto& asset : discovered_assets_for_chain_clone) {
+    discovered_assets_for_chain_clone.push_back(asset.Clone());
+  }
+  std::move(callback).Run(chain_id,
+                          std::move(discovered_assets_for_chain_clone),
+                          std::move(error), error_message);
+
+  // Do not emit event or modify remaining_chains_ count if DiscoverAssets call
+  // was triggered by an AccountsAdded event
+  if (triggered_by_accounts_added) {
+    return;
+  }
+
+  // Complete the call by decrementing remaining_chains_, storing the discoverd
+  // assets for later, and emitting the event if this was the final chain to
+  // finish
+  remaining_chains_--;
+  for (const auto& asset : discovered_assets_for_chain) {
+    discovered_assets_.push_back(asset.Clone());
+  }
```

Comment:
Seems no need for a clone, we should be able to std::move those entries directly.

Reply:
https://github.com/brave/brave-core/pull/15707#discussion_r1026652965 ++